File: solutions/clojure/square-root/1/src/square_root.clj
(ns square-root)

(defn square-root
  "Calculates a number's square root"
  [n]
  (->> n
       (iterate #(float (/ (+ % (/ n %)) 2)))
       (partition 2 1)
       (take-while (fn [[a b]] (not= a b)))
       (last)
       (first)
       int))

File: solutions/python/square-root/1/square_root.py
"""Calculate integer square roots."""


def square_root(number):
    """Return the whole-number square root of number."""
    if number == 1:
        return 1

    low = 1
    high = number // 2

    while low <= high:
        candidate = (low + high) // 2
        square = candidate * candidate

        if square == number:
            return candidate

        if square < number:
            low = candidate + 1
        else:
            high = candidate - 1

    raise ValueError("Number does not have a whole-number square root.")
